text/0000-global-packages.md

```diff
@@ -0,0 +1,195 @@
+2. Detect binaries and write config files _after_ a successful install.
+
+For installs run with `volta install`, we can do the same and directly call `npm i -g` as if the user had called it themselves.
+
+### Node Version
+
+Currently, the Node version that we pin to a package is determined by the `engines` property in the package's `package.json`, finding the latest LTS version of Node that meets the requirements (or the current LTS version if no `engines` is specified). However, that results in the choice of Node being opaque and not directly controllable. Instead, we should default to the user's current default Node, so that the tool is installed in an environment that matches the user's current default configuration.
+
+This has the benefit of providing a clear way to upgrade the pinned Node if a user desires: Switch their default and re-install the package. While that looks a bit like other package managers—which require you to re-install packages for every different version of Node—the key difference is that with Volta, this path is entirely _optional_ and not required unless you want to change the version of Node associated with a package.
+
+### Handling Multiple Installs
+
+If the user calls `npm i -g <package_1> <package_2>`, that would normally all run as a single command. However, to maintain Volta's Sandbox-per-package behavior, we will need to split that out and run `npm i -g <package_1>` and `npm i -g <package_2>` separately, allowing us to install them in separate locations and avoid combining them into a single block.
+
+### Note on `pnpm`
+
+`pnpm` can also be redirected with a `prefix` configuration setting, however we also need to do some messing with the PATH when calling `pnpm i -g`. The reason is that `pnpm` searches the `PATH` to find the highest-precedence "appropriate" location to install the binary symlinks. Unfortunately, that currently means in the Volta image directory for the specific Node version, which is definitely _not_ where we want them. We will likely need to create a directory and add it to the PATH for `pnpm` global installs, to make sure that the symlinks are created in the correct location.
+
+## 2. Shared Global Package Directory
+
+In order to satisfy constraints 4 & 5, allowing global packages to `require` each other, we should create a single shared "globals" directory that we add to `NODE_PATH` when executing a global binary (and _only_ for global binaries, as normal Node commands wouldn't have access the globals anyway). That directory can have symlinks to all of the individual package directories, so that a `require` from any one of the packages will be able to access any of the others. Per the [Node documentation](https://nodejs.org/api/modules.html#modules_loading_from_the_global_folders), the `NODE_PATH` will only be searched if the regular filesystem process fails, so this should result in the correct precedence when resolving a global library.
```

Comment:
Is this `"globals"` directory truly just one for all of Volta, or is it one per Node installation? If package `foo` is installed bound to Node 12 and package `bar` is installed bound to Node 14, calling into one another might not work at all (for example if `foo` has native components compiled to the Node 12 ABI).

Reply:
The thought was to have a single directory, which potentially could cause problems of packages across Node versions not working.

The flip side of that, however, is if we isolate by Node version then there’s the potential for confusion when `require` fails to find a package that _was_ installed.

My sense is that 1) native modules are relatively rare and 2) Default versions change infrequently enough to minimize the potential breakages.

Another concern is even if we isolate packages by Node major version, there’s no guarantee it’ll work across versions (different features / bug fixes within a given major version). So there is always going to be the potential for breakage that we’ll need to deal with.

Reply:
It seems worth calling out somewhere that we expect native modules to be rare in global packages, and that this means that if people _do_ happen to have Node-version-sensitive global modules, they'll need to be careful about what version of Node they're using. That seems reasonable to me, it's just worth being explicit about it in the RFC.

Also, note when it comes to our (separate) work on `npm link` etc, we may not be able to make the same assumption. In local packages, native modules are more common. Not relevant for this RFC, I just want to make sure we don't over-learn from the assumptions we make here and apply them elsewhere where they may not hold.

Your reasoning here sounds plausible to me but should it be made more explicitly in the RFC?

Reply:
Added a "Critique" section covering that and noting the need to clearly document the potential for incompatibility and how to fix it. Might be worth creating a "Troubleshooting" or something section in the docs to cover FAQ type questions and known issues.

Reply:
Also definitely agree that `npm link` definitely needs to be more explicit about the Node versions. Luckily in that case (at least if both projects are using Volta), we can detect a mismatch and show a warning at the time the `link` is created.

Reply:
We may be able to detect when a user is installing a package that contains native modules, and give some kind of appropriate warning: https://stackoverflow.com/questions/36032250/how-to-find-native-modules-in-a-node-js-dependency-tree

Reply:
Oh great find @mikrostew! I’ll add a note that we should show a note when installing a native module to remind the user that they should be aware of Node version compatibility